### PR TITLE
Add JWT authentication with security configuration

### PR DESCRIPTION
**Pull request by SkillfulDev**
*  Add security filter chain
*  Add Jackson config
*  Add JWT autenticaton classes
*  Add custom user details service
*  Add hibernate SQL style 
*  Add role entity

### CODE REVIEW

File: src/main/java/online/carsharing/entity/Role.java [hunk 1]
```diff
@@ -1,6 +1,37 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false, unique = true)
+    private RoleName name;
+
+    @Override
+    public String getAuthority() {
+        return "ROLE_" + name.name();
+    }
```
Comment (HAIOVYI) [inline src/main/java/online/carsharing/entity/Role.java]:
```suggestion
    private RoleType type;

    @Override
    public String getAuthority() {
        return "ROLE_" + type.name();
    }
```

File: src/main/resources/application.properties [hunk 2]
```diff
@@ -9,3 +9,9 @@ spring.jpa.show-sql=true
 spring.jpa.open-in-view=false
 
 server.servlet.context-path=/api
+
+spring.jpa.properties.hibernate.format_sql=true
+spring.jpa.properties.hibernate.highlight_sql=true
+
+jwt.secret.key = ${JWT_SECRET_KEY}
+jwt.expetarion.time = 50 000 000
```
Comment (azzilian) [inline src/main/resources/application.properties]: Mb it should be Environment Variable as well

**Review by azzilian — CHANGES_REQUESTED** [review]
lets add EV to expiration date for resources just in case to not forget later

File: src/main/java/online/carsharing/repository/user/RoleRepository.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package online.carsharing.repository.user;
+
+import java.util.Optional;
+import online.carsharing.entity.Role;
+import online.carsharing.entity.RoleName;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    Optional<Role> findByName(RoleName roleName);
```
Comment (HAIOVYI) [inline src/main/java/online/carsharing/repository/user/RoleRepository.java]:
```suggestion
    Optional<Role> findByName(RoleType roleType);
```

File: src/main/java/online/carsharing/security/JwtUtil.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    public boolean isValidToken(String token) {
+        try {
+            Jws<Claims> claimsJws = Jwts.parserBuilder()
+                    .setSigningKey(secretKey)
+                    .build()
+                    .parseClaimsJws(token);
+
+            return !claimsJws.getBody().getExpiration().before(new Date());
+        } catch (JwtException e) {
+            throw new JwtTokenValidException("Expired or invalid JWT token", e);
```
Comment (HAIOVYI) [inline src/main/java/online/carsharing/security/JwtUtil.java:44]:
```suggestion
            return false;
```

File: src/main/java/online/carsharing/security/impl/AuthenticationServiceImpl.java [hunk 5]
```diff
@@ -31,22 +42,25 @@ public UserResponseDto register(UserRegisterRequestDto registerDto) {
                     + registerDto.getEmail() + " already exists");
         }
         User user = userMapper.toUser(registerDto);
-        user.setRole(Role.CUSTOMER);
+        Role role = roleRepository.findByName(RoleName.CUSTOMER).orElseThrow(() ->
+                new NoSuchElementException("Role CUSTOMER not found"));
```
Comment (HAIOVYI) [inline src/main/java/online/carsharing/security/impl/AuthenticationServiceImpl.java]: `@PostConstruct
`

**Review by HAIOVYI — CHANGES_REQUESTED** [review]
GJ!

**Review by azzilian — DISMISSED** [review]
Gj! Aprrove!

**Review by HAIOVYI — APPROVED** [review]
Very good!

**Review by azzilian — APPROVED** [review]

**Pull request merged**